### PR TITLE
Feat/world of fantasy

### PR DESCRIPTION
**Pull request by Sotatek-Bean**
feat/world-of-fantasy and main branch are created after code is done in "Master"

Just for easier to review changes from "https://angular.io/tutorial/toh-pt5",

The better version to try is from branch "master"

### CODE REVIEW

File: README.MD [hunk 1]
```diff
@@ -0,0 +1,33 @@
+For eaiser to understand, i made 1 demo video here:
+https://drive.google.com/file/d/1M74bweDVCPdTU-3Y96wGQHs41YvEtfBL/view?usp=sharing
```
Comment (ChrisPriebe) [inline README.MD:2]: Great video, very helpful

File: README.MD [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+  When defeat enemy, they get Money and sometime drop Items.
+
+  Level up, Find better Equipment then become the top leader of Adventurers
+
+Time used: Plan/idea structure designed 1-2 hours, almost 4 days coding with 2-3 hours each.
+
+Total: around 12-15 hours.
+
+Improvements later:
```
Comment (ChrisPriebe) [inline README.MD:17]: Missing
 - Installation instructions
 - Test coverage
 - architecture designs and decisions

File: e2e/src/app.e2e-spec.ts [hunk 3]
```diff
@@ -0,0 +1,170 @@
+import { browser, element, by, ElementFinder } from 'protractor';
+
+const expectedH1 = 'Tour of Heroes';
+const expectedTitle = `${expectedH1}`;
+const targetHero = { id: 16, name: 'RubberMan' };
+const targetHeroDashboardIndex = 3;
+const nameSuffix = 'X';
+const newHeroName = targetHero.name + nameSuffix;
+
```
Comment (ChrisPriebe) [inline e2e/src/app.e2e-spec.ts:9]: I was glad to see end to end tests.  What is the command to turn it

Conventially system consts should be all caps EXPECTED_H1 but that is open to debate on many projects, our lint step will standardize it for our project

File: e2e/src/app.e2e-spec.ts [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+    return new Hero(
+      +s.substring(0, s.indexOf(' ')),
+      s.slice(s.indexOf(' ') + 1),
+    );
+  }
+
+  // Get hero id and name from the given detail element.
+  static async fromDetail(detail: ElementFinder): Promise<Hero> {
+    // Get hero id from the first <div>
+    const id = await detail.all(by.css('div')).first().getText();
```
Comment (ChrisPriebe) [inline e2e/src/app.e2e-spec.ts:26]: It would be better to use an CSS id or class b/c as the HTML changes the tests will fail but the ID will more often stay the same.

File: e2e/src/app.e2e-spec.ts [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+      navElts,
+
+      appDashboardHref: navElts.get(0),
+      appDashboard: element(by.css('app-root app-dashboard')),
+      topHeroes: element.all(by.css('app-root app-dashboard > div a')),
+
+      appHeroesHref: navElts.get(1),
+      appHeroes: element(by.css('app-root app-heroes')),
+      allHeroes: element.all(by.css('app-root app-heroes li')),
+      heroDetail: element(by.css('app-root app-hero-detail > div'))
```
Comment (ChrisPriebe) [inline e2e/src/app.e2e-spec.ts:53]: Using a class name would be more future friendly here.

File: e2e/src/app.e2e-spec.ts [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+
+  describe('Initial page', () => {
+
+    it(`has title '${expectedTitle}'`, async () => {
+        expect(await browser.getTitle()).toEqual(expectedTitle);
+    });
+
+    it(`has h1 '${expectedH1}'`, async () => {
+      await expectHeading(1, expectedH1);
+    });
```
Comment (ChrisPriebe) [inline e2e/src/app.e2e-spec.ts:65]: These two cases are fairly simplistic.  It is better to use tests to cover logic, especially logic that could have edge cases.  In this cases it is almost testing if HTML works but it also preventing fixes like when we want to update the title.  Such a change is not based on "logic" but style which is not really the best for e2e tests

File: e2e/src/app.e2e-spec.ts [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+    beforeAll(() => browser.get(''));
+
+    it('has top heroes', async () => {
+      const page = getPageElts();
+      expect(await page.topHeroes.count()).toEqual(4);
+    });
+
+    it(`selects and routes to ${targetHero.name} details`, dashboardSelectTargetHero);
+
+    it(`updates hero name (${newHeroName}) in details view`, updateHeroNameInDetailView);
```
Comment (ChrisPriebe) [inline e2e/src/app.e2e-spec.ts:91]: This is a better test as now we are testing functionality

File: e2e/src/app.e2e-spec.ts [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+
+    it(`saves and shows ${newHeroName} in Dashboard`, async () => {
+      await element(by.buttonText('go back')).click();
+      const targetHeroElt = getPageElts().topHeroes.get(targetHeroDashboardIndex);
+      expect(await targetHeroElt.getText()).toEqual(newHeroName);
+    });
```
Comment (ChrisPriebe) [inline e2e/src/app.e2e-spec.ts:97]: Testing it saves and restores is great. 
Typcially however a test should be a single task like save or show name, not both

File: e2e/src/app.e2e-spec.ts [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+    const targetHeroElt = getPageElts().topHeroes.get(targetHeroDashboardIndex);
+    expect(await targetHeroElt.getText()).toEqual(targetHero.name);
+    await targetHeroElt.click();
+
+    const page = getPageElts();
+    expect(await page.heroDetail.isPresent()).toBeTruthy('shows hero detail');
+    const hero = await Hero.fromDetail(page.heroDetail);
+    expect(hero.id).toEqual(targetHero.id);
+    expect(hero.name).toEqual(targetHero.name.toUpperCase());
+  }
```
Comment (ChrisPriebe) [inline e2e/src/app.e2e-spec.ts:142]: The tests are missing your core functionality of the fight,
Also missing that on mobile there is no click event but a touch so mobile does not work to start a fight

File: src/app/app.component.ts [hunk 10]
```diff
@@ -1,10 +1,19 @@
...
+import { Component, OnInit, inject } from '@angular/core';
+import { RouterModule } from '@angular/router';
+import { UserService } from './services/user.service';
 
 @Component({
   selector: 'app-root',
   templateUrl: './app.component.html',
-  styleUrls: ['./app.component.css']
+  styleUrls: ['./app.component.css'],
+  standalone: true,
```
Comment (ChrisPriebe) [inline src/app/app.component.ts:10]: It's nice to see this using the latest Angular conventions

File: src/app/pages/components/hero-detail/hero-detail.component.css [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+}
+input {
+  font-size: 1em;
+  padding: .5rem;
+}
+.card {
+  max-width: 500px;
+  img {
+    width: 100%;
+  }
```
Comment (ChrisPriebe) [inline src/app/pages/components/hero-detail/hero-detail.component.css:13]: What about responsive layouts

File: src/app/pages/components/hero-detail/hero-detail.component.html [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+    <h2>{{hero.name | uppercase}}</h2>
+    <div>
+      <label for="hero-name">Hero name: </label>
+      <input id="hero-name" [(ngModel)]="hero.name" placeholder="Hero name"/>
+      <div class="stat-details d-flex justify-content-between mt-1">
+        @if (entityService.getHeroFightStats(this.hero.id) | async; as fightStats) {
+          <span>Atk: {{fightStats.atk}}</span>
+          <span>Speed: {{fightStats.speed}}</span>
+          <span>HP: {{fightStats.health}}</span>
+          <span>Level: {{hero.level}}</span>
```
Comment (ChrisPriebe) [inline src/app/pages/components/hero-detail/hero-detail.component.html:12]: It's good practice to add a class tag to the span

File: src/app/pages/components/inventory/inventory.component.scss [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+      margin: 10px 0 0 0;
+    }
+    img {
+      max-width: 100%;
+    }
+  }
+  .equiped {
+    box-shadow:0 5px 10px 5px rgba(129, 179, 139, 0.76);
+  }
+}
```
Comment (ChrisPriebe) [inline src/app/pages/components/inventory/inventory.component.scss:23]: this works but it may have been easier to use material or ionic and had a lot of built in layout pieces.

File: src/app/pages/components/inventory/inventory.component.ts [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+    this.entityService.getArmors().pipe(takeUntil(this.unsub$))
+    .subscribe(armors => this.armors = armors);
+    this.entityService.getWeapons().pipe(takeUntil(this.unsub$))
+    .subscribe(weapons => this.weapons = weapons);
+  }
+
+  // equip item to hero
+  equipItem(item: Item) {
+    if (this.hero) {
+      this.hero.equip[item.type] = this.hero.equip[item.type] === item.id ? undefined : item.id;
```
Comment (ChrisPriebe) [inline src/app/pages/components/inventory/inventory.component.ts:44]: Might be better abstracted as a hero class that is OOP, then hero.equip() allowing for separation of concerns and future friendly changes.

File: src/app/pages/components/inventory/inventory.component.ts [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+    }
+  }
+
+  sellItem(item: Item) {
+    this.entityService.removeItem(item);
+    this.userService.addMoney(item.level * 10)
```
Comment (ChrisPriebe) [inline src/app/pages/components/inventory/inventory.component.ts:50]: good abstraction with the entity and user service.

File: src/app/pages/dashboard/dashboard.component.html [hunk 16]
```diff
@@ -0,0 +1,7 @@
+<div class="heroes-menu">
+  <a *ngFor="let hero of heroes" (click)="setHero(hero)">
```
Comment (ChrisPriebe) [inline src/app/pages/dashboard/dashboard.component.html:2]: Angular 17 switches to @for

File: src/app/pages/dashboard/dashboard.component.ts [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+  getHeroes(): void {
+    this.entityService.getHeroes(true).pipe(takeUntil(this.unsub$))
+      .subscribe(heroes => this.heroes = heroes);
+  }
+
+  async setHero(hero: Hero) {
+    if (!this.playService.paused) {
+      return;
+    }
+    const heroFightStats = {...hero, ...(await firstValueFrom(this.entityService.getHeroFightStats(hero.id)))};
```
Comment (ChrisPriebe) [inline src/app/pages/dashboard/dashboard.component.ts:45]: firstValueFrom - Thank-you, I learnt something new, been looking for a way to do that

File: src/app/services/canvas.service.ts [hunk 18]
```diff
@@ -0,0 +1,276 @@
...
+      const animation = this.startAnimation(this.heroGroup, AnimationType.heal, 100, 300);
+      setTimeout(() => {
+        animation.stop();
+        this.heroGroup.x(100);
+        this.heroGroup.y(350);
+        playService.doAction(Actions.heal);
+      }, 1000);
+      actionGroup.hide();
+    });
+    attackGroup.on('click', () => {
```
Comment (ChrisPriebe) [inline src/app/services/canvas.service.ts:77]: mobile: tap or touch ?? but not click

File: src/app/services/canvas.service.ts [hunk 19]
```diff
@@ -0,0 +1,276 @@
...
+  initMonster(monster: Monster) {
+    this.monsterGroup.destroy();
+    this.monsterGroup = new Konva.Group({
+      x: 700,
+      y: 200,
+    });
+    this.monsterGroup.add(this.createImageLayer(0, 0, 250, 250, monster.avatar || monster.name));
+    this.initMonsterInfo(monster);
+    this.layer.add(this.monsterGroup);
+  }
```
Comment (ChrisPriebe) [inline src/app/services/canvas.service.ts:204]: Better abstraction to add these as their own classes (but it is listed in the README so you already caught this which makes me glad)

**Review by ChrisPriebe — COMMENTED** [review]
Thank-you for working on this challenge.  I have left some comments as a chance to grow in your coding skill.   I liked the additions of the upgrade hero, armour and weapons.  I also liked the animations and how fast you where able to achieve it.

File: e2e/src/app.e2e-spec.ts (hunk 3 above)
Comment (Sotatek-Bean) [reply e2e/src/app.e2e-spec.ts:9]: it was ng e2e
But these are tests for base project so it might not run properly now
